### ISSUE TITLE
Cache modify API for adding fields

### ISSUE THREAD

**Issue by danReynolds** [apollographql/apollo-client#6922]
I noticed that the cache#modify API can only modify existing fields on a store object as iterated here: https://github.com/apollographql/apollo-client/blob/main/src/cache/inmemory/entityStore.ts#L141. Was it a design choice to limit its scope to only modifying existing fields as opposed to adding new data? I can add new fields with writeFragment but I would prefer the modify API instead of having to do a read/write pattern.

**Comment by i-am-chris-ho** [comment]
According to the current docs: https://www.apollographql.com/docs/react/local-state/managing-state-with-field-policies/#modifying

> If you're using the cache directly, call one of writeQuery, writeFragment, or cache.modify (all documented here) to modify cached fields. Like reactive variables, all of these methods trigger a refresh of every affected active operation.

It seems that it was not a design choice. Or it may be at odds with the current local-state documentation.

**Comment by danReynolds** [comment]
> According to the current docs: https://www.apollographql.com/docs/react/local-state/managing-state-with-field-policies/#modifying
> 
> > If you're using the cache directly, call one of writeQuery, writeFragment, or cache.modify (all documented here) to modify cached fields. Like reactive variables, all of these methods trigger a refresh of every affected active operation.
> 
> It seems that it was not a design choice. Or it may be at odds with the current local-state documentation.

Well that says to modify cached fields, which it does, it just cannot introduce new fields like writeFragment can.

**Comment by i-am-chris-ho** [comment]
> Well that says to modify cached fields, which it does, it just cannot introduce new fields like writeFragment can.

True. I guess it is confusing that it is in the local-state/local-only fields section of the docs

**Comment by benjamn** [comment]
This was a deliberate design choice!

When you're adding fields, it's important to go through the whole process of serializing field arguments and running `merge` functions, which `writeQuery` or `writeFragment` do, but `cache.modify` does not.

More generally, a very common criticism of Apollo Client is that there's more than one way to do many things. In this case, we already have `writeQuery` and `writeFragment` (which, to be fair, have slightly different uses), so I am not eager to introduce another tool that covers that use case, especially since `cache.modify` doesn't really handle everything that needs to happen, so we'd have to document its shortcomings in detail. I don't think I can explain/motivate that story to someone who isn't already a fan of `cache.modify`.

In short, please accept the narrow role of `cache.modify`, which is to modify existing data.

**Comment by bstro** [comment]
It would be helpful if the documentation mentioned why one might use these three different options; with the existing documentation, its a bit fuzzy to me why I would use one over the other (in one example in [the Mutation docs](https://www.apollographql.com/docs/react/data/mutations/), a writeFragment operation is nested within a call to modify, unclear why).

Moreover, the [documentation states](https://www.apollographql.com/docs/react/data/mutations/#making-all-other-cache-updates) `When the ADD_TODO mutation is run in the above example, the newly added and returned todo object is saved into the cache. The previously cached list of todos, being watched by the GET_TODOS query, is not automatically updated however. This means the GET_TODOS query isn't notified that a new todo was added, which then means the query will not update to show the new todo. To address this we're using cache.modify which gives us a way to surgically insert or delete items from the cache, by running modifier functions.` — but then under [the documentation for writeQuery and writeFragment](https://www.apollographql.com/docs/react/caching/cache-interaction/#writequery-and-writefragment), the docs state that `All subscribers to the Apollo Client cache (including all active queries) see this change and update your application's UI accordingly.`.

So it is ultimately unclear to me how to get my queries to refresh when they depend on data from the cache, the docs seem contradictory.
